kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogReplayEngineMetricsSuite.scala

```diff
@@ -342,6 +379,230 @@ class LogReplayEngineMetricsSuite extends QueryTest
+        // Attempt to read 10.crc fails and read 10.checkpoint.parquet succeeds.
+        expJsonVersionsRead = Nil,
+        expParquetVersionsRead = Seq(10),
+        expParquetReadSetSizes = Seq(1),
+        expChecksumReadSet = Seq(10)
+      )
+    }
+  }
+
+  test(
+    "checksum missing read version and the previous version, " +
+    "checkpoint exists the read version the previous version, use checkpoint"
+  ) {
+    withTempDirAndMetricsEngine { (path, engine) =>
+      withSQLConf(DeltaSQLConf.DELTA_WRITE_CHECKSUM_ENABLED.key -> "true") {
+        spark.sql(
+          s"CREATE TABLE delta.`$path` USING DELTA AS " +
+          s"SELECT 0L as id"
+        )
+        for (_ <- 0 to 10) { appendCommit(path) }
+      }
+      val checkpointVersion = 10;
+      assert(
+        Files
+          .exists(new File(path, f"_delta_log/$checkpointVersion%020d.checkpoint.parquet").toPath)
+      )
+      assert(Files.deleteIfExists(new File(path, f"_delta_log/$checkpointVersion%020d.crc").toPath))
+      assert(
+        Files.deleteIfExists(new File(path, f"_delta_log/${checkpointVersion + 1}%020d.crc").toPath)
+      )
+
+      val table = Table.forPath(engine, path)
+
+      // 11.crc, 10.crc missing, 10.checkpoint.parquet exists.
+      // Attempt to read 11.crc fails and read 10.checkpoint.parquet and 11.json succeeds.
+      loadPandMCheckMetrics(
+        table.getSnapshotAsOfVersion(engine, 11).getSchema(engine),
+        engine,
+        expJsonVersionsRead = Seq(11),
+        expParquetVersionsRead = Seq(10),
+        expParquetReadSetSizes = Seq(1),
+        expChecksumReadSet = Seq(11)
+      )
+    }
+  }
+
+  test(
+    "checksum missing read version, " +
+    "both checksum and checkpoint exist the read version the previous version, use checksum"
+  ) {
+    withTempDirAndMetricsEngine { (path, engine) =>
+      withSQLConf(DeltaSQLConf.DELTA_WRITE_CHECKSUM_ENABLED.key -> "true") {
+        spark.sql(
+          s"CREATE TABLE delta.`$path` USING DELTA AS " +
+          s"SELECT 0L as id"
+        )
+        for (_ <- 0 to 10) { appendCommit(path) }
+      }
+      val checkpointVersion = 10;
+      assert(
+        Files
+          .exists(new File(path, f"_delta_log/$checkpointVersion%020d.checkpoint.parquet").toPath)
+      )
+      assert(
+        Files.deleteIfExists(new File(path, f"_delta_log/${checkpointVersion + 1}%020d.crc").toPath)
+      )
+
+      val table = Table.forPath(engine, path)
+
+      // 11.crc, missing, 10.crc and 10.checkpoint.parquet exist.
+      // Attempt to read 11.crc fails and read 10.checkpoint.parquet and 11.json succeeds.
+      loadPandMCheckMetrics(
+        table.getSnapshotAsOfVersion(engine, 11).getSchema(engine),
+        engine,
+        expJsonVersionsRead = Seq(11),
+        expParquetVersionsRead = Nil,
+        expParquetReadSetSizes = Nil,
+        expChecksumReadSet = Seq(11, 10)
+      )
+    }
+  }
+
+  def buildTableWithCrc(path: String): Unit = {
```

Comment:
what are the assumptions of what delta talbe state this method produces? 0.json to 11.json, 0.crc to 11.crc, 10.checkpoint.parquet?

Can we (a) state these assumptions, (b) assert that they are true, and (c) inline in the test succinctly what his has yieled e.g. `buildTableWithCrc(path) // 0 to 11 .json, 0 to 11.crc, 10.checkpoint.parquet`

This will make understanding the tests much easier.

Reply:
>0.json to 11.json, 0.crc to 11.crc, 10.checkpoint.parquet
Yes correct.
Added comments

Reply:
Please verbatim include this comment in the `buildTableWithCrc(path)` usages above

`buildTableWithCrc(path) // 0 to 11 .json, 0 to 11.crc, 10.checkpoint.parquet`